pos-print/src/test/java/com/clouway/pos/print/adapter/http/BackgroundReceiptPrinterServiceTest.kt

```diff
@@ -0,0 +1,126 @@
+package com.clouway.pos.print.adapter.http
+
+import com.clouway.pos.print.adapter.db.BackgroundReceiptPrinterService
+import com.clouway.pos.print.adapter.db.ReceiptRepository
+import com.clouway.pos.print.adapter.db.ReceiptStatus
+import com.clouway.pos.print.core.*
+import com.clouway.pos.print.persistent.DatastoreCleaner
+import com.clouway.pos.print.persistent.DatastoreRule
+import com.clouway.pos.print.printer.Status
+import org.bson.Document
+import org.jmock.AbstractExpectations.returnValue
+import org.jmock.AbstractExpectations.throwException
+import org.jmock.Expectations
+import org.jmock.Mockery
+import org.jmock.integration.junit4.JUnitRuleMockery
+import org.junit.ClassRule
+import org.junit.Rule
+import org.junit.Test
+
+/**
+ * @author Tsvetozar Bonev (tsvetozar.bonev@clouway.com)
+ */
+class BackgroundReceiptPrinterServiceTest {
+
+  @Rule
+  @JvmField
+  var context = JUnitRuleMockery()
+
+  companion object {
+    @ClassRule
+    @JvmField
+    val dataStoreRule = DatastoreRule()
+  }
+
+  @Rule
+  @JvmField
+  var cleaner = DatastoreCleaner(dataStoreRule.db())
+
+
+  private fun Mockery.expecting(block: Expectations.() -> Unit) {
+    checking(Expectations().apply(block))
+  }
+
+  private val factory = context.mock(PrinterFactory::class.java)
+  private val printer = context.mock(ReceiptPrinter::class.java)
+  private val repo = context.mock(ReceiptRepository::class.java)
+  private val notifier = context.mock(PrintNotifier::class.java)
+
+  private val service = BackgroundReceiptPrinterService(repo, factory, notifier)
+
+  private val collection = dataStoreRule.db()!!.getCollection("test")
+
+  private val receipt = Receipt.newReceipt()
+    .withReceiptId("::receiptId::")
+    .withAmount(200.0)
+    .build()
+
+  private val receiptDoc = Document()
+    .append("receiptId", "::receiptId::")
+    .append("amount", 200.0)
+    .append("prefixLines", emptyList<String>())
+    .append("suffixLines", emptyList<String>())
+    .append("currency", "::currency::")
+    .append("receiptItems", emptyList<Document>())
+    .append("status", ReceiptStatus.PRINTING.name)
+    .append("sourceIp", "::sourceIp::")
+    .append("isFiscal", true)
+
+  private val acceptedPrintingResponse = PrintReceiptResponse(setOf(Status.FISCAL_RECEIPT_IS_OPEN))
+  private val rejectedPrintingResponse = PrintReceiptResponse(setOf(Status.BROKEN_PRINTIN_MECHANISM))
+
+  @Test
+  fun printReceipt() {
+    context.expecting {
+      oneOf(factory).getPrinter("::sourceIp::")
+      will(returnValue(printer))
+
+      oneOf(printer).printFiscalReceipt(receipt)
+      will(returnValue(acceptedPrintingResponse))
+
+      oneOf(repo).finishPrinting("::receiptId::")
+      will(returnValue(receipt))
+
+      oneOf(notifier).notifyPrinted(acceptedPrintingResponse)
+    }
+
+    collection.insertOne(receiptDoc)
```

Comment:
Looks really strange to be tested like that as the test knows the structure of the receipt in the queue. The test shouldn't know this and should deal with Receipt object or ReceiptScheduledForPrinting.

Reply:
Separated mongo logic from the background thread's dependencies.